operator/rpc_client.go

```diff
@@ -42,82 +75,172 @@ func (c *AggregatorRpcClient) dialAggregatorRpcClient() error {
 	return nil
 }
 
-func (c *AggregatorRpcClient) sendRequest(sendCb func() error, maxRetries int, retryInterval time.Duration) {
-	if c.rpcClient == nil {
-		c.logger.Info("rpc client is nil. Dialing aggregator rpc client")
-		err := c.dialAggregatorRpcClient()
+func (c *AggregatorRpcClient) InitializeClientIfNotExist() error {
+	c.rpcClientMutex.Lock()
+	defer c.rpcClientMutex.Unlock()
+
+	if c.rpcClient != nil {
+		return nil
+	}
+
+	return c.dialAggregatorRpcClient()
+}
+func (c *AggregatorRpcClient) onTick() {
+	tickerC := c.resendTicker.C
+	for {
+		// TODO(edwin): handle closed chan
+		<-tickerC
+
+		// Critically ugly section
+		{
+			c.unsentMessagesMutex.Lock()
+			if len(c.unsentMessages) == 0 {
+				c.unsentMessagesMutex.Unlock()
+				continue
+			}
+			c.unsentMessagesMutex.Unlock()
+		}
```

Comment:
```suggestion
		c.unsentMessagesMutex.Lock()
		unsentMessagesLen := len(c.unsentMessages)
		c.unsentMessagesMutex.Unlock()

		if unsentMessagesLen == 0 {
			continue
		}
```

Reply:
Technically you risking of having actual `len` changed since its not in critical section anymore and someone else can jump in between

Reply:
Those seem like equivalent approaches for me, though of course depends on what it turns into when compiling. How do you think they're differing?

Reply:
I'm not sure we can assume it's critical session once the mutex is unlocked

Reply:
The thing is they in proposed approach you're not checking actual `len` of array, you checking `len` that was in critical section - snapshot at that moment. Since there's no lock anymore, another routine could take over, change `c.unsentMessages` and you would basing your decisions on incorrect `len`(`unsentMessagesLen `).

This is wrong way to check if array size is 0 in that particular moment

Reply:
Yup, but both have the same problem. Again, the only operation that is relevant in terms of the critical session itself is getting the length. Concurrency can only affect the length, not the check. If immediately after the lock is released in the current implementation it was changed, the outcome would be exactly the same.

Reply:
Btw, can't we simply start and stop the ticker to avoid useless iterations and avoid this double check altogether? This is already done on `tryResend` anyway.

Reply:
Anyway, if we can't we can do any of the two approaches, np. In my opinion both work separately, but the truth is it's a workaround to avoid re-initializing the client.
Another option would be just doing it inside `tryResend` after the length check, seems better even.

Reply:
If we are to keep the current version somehow, I do think the comment is unnecessary

Reply:
> he only operation that is relevant in terms of the critical session itself is getting the length

The critical section here is making decision if we need to skip or not and for that you need a correct length.

This check exist for not triggering client initilization prematurely since it shall be done only on in first send.

> Concurrency can only affect the length, not the check

check depend on the length. In critical section its valid, outside not.
current way:
If we didn't `continue` in critical section then we don't care what current length is in context of initializing client. If it wasn't 0 we can try initializing.

proposed one.
The value in critical section was 0. Current one unknown. Basing skip on past

Reply:
Keeping track of internal conversations - we'll keep the current approach for this PR and I'll just push another PR later removing this check and centralizing the RPC client initialization on `tryResendFromDeque`.